src/hpke/registry.rs

```diff
@@ -41,12 +42,56 @@ impl KeyPair {
     }
 }
 
+// This newtype is necessary because IpaPublicKey is an associated type from another crate (hpke).
+// The coherence rules prohibit us from implementing `PublicKeyRegistry` both for our concrete type
+// `KeyPair` and for `IpaPublicKey`, because the impls would overlap if hpke chose to define
+// `IpaPublicKey` to be the same as `KeyPair`.
+pub struct PublicKeyOnly(pub IpaPublicKey);
+
+impl Deref for PublicKeyOnly {
+    type Target = IpaPublicKey;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+pub trait PublicKeyRegistry {
```

Comment:
One nice thing about the key registry abstraction (even for managing public keys) is that it provides some assurance the correct key ID is used. A seal API that passes key ID and `&key` directly has more risk that the passed key ID doesn't match the passed key.

However it may make sense to have the lowest level seal API take key ID and key separately, and then provide higher level wrappers for particular use cases. We may want to add key IDs to `NetworkConfig` and then we could provide a seal API that takes `NetworkConfig` rather than key IDs / keys.

Reply:
agree, I wouldn't want to pass both key id and `&key` -  that defeats the purpose. I am on the same page wrt to configuration, although I might prefer a wrapper around `NetworkConfig` to work with encryption - but essentially it will use just client configuration to encrypt